File: runtime/go/dis.go
package shani

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"
	"time"
)

// DISStateMachine governs Shani's integrity posture (SPEC §4.4).
//
// Transitions are explicit and logged. VIOLATED → VALID requires
// explicit reset with documented justification.
type DISStateMachine struct {
	state   DIS
	history []DisTransition
}

// NewDISStateMachine creates a new machine in VALID state.
func NewDISStateMachine() *DISStateMachine {
	return &DISStateMachine{state: DISValid}
}

// State returns the current DIS state.
func (m *DISStateMachine) State() DIS {
	return m.state
}

// History returns a copy of all recorded transitions.
func (m *DISStateMachine) History() []DisTransition {
	result := make([]DisTransition, len(m.history))
	copy(result, m.history)
	return result
}

// Transition attempts a state change. VIOLATED → VALID must use ResetToValid.
func (m *DISStateMachine) Transition(to DIS, reason, triggeredBy string) (DisTransition, error) {
	if m.state == DISViolated && to == DISValid {
		return DisTransition{}, fmt.Errorf(
			"use ResetToValid() with justification to recover from VIOLATED state",
		)
	}

	record := DisTransition{
		FromState:   m.state,
		ToState:     to,
		Reason:      reason,
		Timestamp:   time.Now().UTC(),
		TriggeredBy: triggeredBy,
	}
	m.history = append(m.history, record)
	m.state = to
	logTransition(record)
	return record, nil
}

// ResetToValid resets DIS from VIOLATED to VALID (SPEC §4.4).
//
// Requires non-empty justification and named human authority.
func (m *DISStateMachine) ResetToValid(justification, authorizedBy string) (DisTransition, error) {
	if m.state != DISViolated {
		return DisTransition{}, fmt.Errorf(
			"ResetToValid() may only be called from VIOLATED state; current state: %s", m.state,
		)
	}
	if strings.TrimSpace(justification) == "" {
		return DisTransition{}, fmt.Errorf("justification must not be empty")
	}
	if strings.TrimSpace(authorizedBy) == "" {
		return DisTransition{}, fmt.Errorf("authorizedBy must name a human authority")
	}

	record := DisTransition{
		FromState:   DISViolated,
		ToState:     DISValid,
		Reason:      "MANUAL RESET — " + justification,
		Timestamp:   time.Now().UTC(),
		TriggeredBy: authorizedBy,
	}
	m.history = append(m.history, record)
	m.state = DISValid
	logTransition(record)
	return record, nil
}

func logTransition(r DisTransition) {
	b, _ := json.Marshal(r)
	slog.Info("DIS transition", "entry", string(b))
}

File: runtime/rust/src/dis.rs
//! DIS State Machine (SPEC §4.4)

use chrono::Utc;
use crate::schemas::state::{Dis, DisTransition};
use crate::error::ShaniError;

/// The DIS State Machine governs Shani's integrity posture.
///
/// Transitions are explicit and logged. VIOLATED → VALID requires
/// explicit reset with documented justification (SPEC §4.4).
pub struct DisStateMachine {
    state: Dis,
    history: Vec<DisTransition>,
}

impl DisStateMachine {
    pub fn new() -> Self {
        Self {
            state: Dis::Valid,
            history: vec![],
        }
    }

    pub fn state(&self) -> Dis {
        self.state
    }

    pub fn history(&self) -> &[DisTransition] {
        &self.history
    }

    /// Attempt a state transition. VIOLATED → VALID must use reset_to_valid().
    pub fn transition(
        &mut self,
        to: Dis,
        reason: impl Into<String>,
        triggered_by: impl Into<String>,
    ) -> Result<DisTransition, ShaniError> {
        if self.state == Dis::Violated && to == Dis::Valid {
            return Err(ShaniError::DelegationViolation(
                "Use reset_to_valid() with justification to recover from VIOLATED".to_string(),
            ));
        }

        let record = DisTransition {
            from_state: self.state,
            to_state: to,
            reason: reason.into(),
            timestamp: Utc::now().to_rfc3339(),
            triggered_by: triggered_by.into(),
        };
        self.history.push(record.clone());
        self.state = to;

        tracing_log(&record);
        Ok(record)
    }

    /// Reset DIS from VIOLATED to VALID (SPEC §4.4).
    ///
    /// Requires non-empty justification and named human authority.
    pub fn reset_to_valid(
        &mut self,
        justification: impl Into<String>,
        authorized_by: impl Into<String>,
    ) -> Result<DisTransition, ShaniError> {
        if self.state != Dis::Violated {
            return Err(ShaniError::DelegationViolation(format!(
                "reset_to_valid() may only be called from VIOLATED state. Current: {:?}",
                self.state
            )));
        }

        let justification = justification.into();
        let authorized_by = authorized_by.into();

        if justification.trim().is_empty() {
            return Err(ShaniError::DelegationViolation("justification must not be empty".into()));
        }
        if authorized_by.trim().is_empty() {
            return Err(ShaniError::DelegationViolation("authorized_by must name a human".into()));
        }

        let record = DisTransition {
            from_state: Dis::Violated,
            to_state: Dis::Valid,
            reason: format!("MANUAL RESET — {justification}"),
            timestamp: Utc::now().to_rfc3339(),
            triggered_by: authorized_by,
        };
        self.history.push(record.clone());
        self.state = Dis::Valid;

        tracing_log(&record);
        Ok(record)
    }
}

impl Default for DisStateMachine {
    fn default() -> Self {
        Self::new()
    }
}

fn tracing_log(record: &DisTransition) {
    eprintln!(
        "[shani.dis.audit] DIS transition: {:?} → {:?} reason='{}' by='{}'",
        record.from_state, record.to_state, record.reason, record.triggered_by
    );
}
